### PR TITLE
Enable XplatEventLogger on all non-Windows platforms

### PR DESCRIPTION
**Pull request by manu-silicon**
Ensure that even if FEATURE_EVENTSOURCE_XPLAT is not defined it, the
implementation provides a dummy implementation. This is required because
mscorlib defines FEATURE_EVENTSOURCE_XPLAT even if this is not supported
by the underlying platform.

### CODE REVIEW

**Comment by jkotas** [comment]
@manu-silicon Could you please take a look at the OSX build break? Looks like the condition for `XplatEventLogger.cs` in `mscorlib.shared.sources.props` needs to be fixed up.

**Comment by manu-silicon** [comment]
@jkotas Just pushed a change to always include ```XplatEventLogger.cs```.

**Comment by jkotas** [comment]
LGTM. Could you please squash into single commit?

**Comment by manu-silicon** [comment]
Squashed.

**Comment by ramarag** [comment]
Hey instead of this big change, you can just unconditionally turn on FEATURE_EVENTSOURCE_XPLAT in the VM, 
currently it is on only if FEATURE_EVENT_TRACE is specified look at https://github.com/dotnet/coreclr/blob/master/CMakeLists.txt#L786

**Comment by ramarag** [comment]
Correct me if i am wrong, this would not have raised if we had ways to trun off features for different flavors of linux in mscorlib

**Comment by manu-silicon** [comment]
The issue was that on Linux ARM mscorlib assumed that Xplat was available but the native part had been disabled due to poor support (see #2824). And indeed if we could have turned on/off some features in mscorlib, it would have matched the native part and there would not have been an issue.

**Comment by ramarag** [comment]
That is exactly the above mentioned solution would do, instead of matching mscorlib to native, we would be changing native to match mscorlib ,only for this instance
so
    if(CLR_CMAKE_PLATFORM_UNIX)
        add_definitions(-DFEATURE_EVENTSOURCE_XPLAT=1)
    endif(CLR_CMAKE_PLATFORM_UNIX)

needs to be moved out of
 if(FEATURE_EVENT_TRACE)

**Comment by jkotas** [comment]
Yes, that would work too - but it would required keeping the managed and unmanaged settings in sync somehow.

The typical approach taken in corefx is to build the managed code general purpose, and have the fine-grained differences in unmanaged C++ wrappers only. Here is one example from many: https://github.com/dotnet/corefx/blob/1db0c74f3f525365c2573458aa1f454cacf00453/src/Native/System.Security.Cryptography.Native/pal_ssl.cpp#L49. 

This corefx approach make it easier to maintain different configurations. It is why I have suggested it here as well.

**Comment by jkotas** [comment]
I see - you are basically suggesting to use `FEATURE_EVENTSOURCE_XPLAT` instead of `FEATURE_PAL` to ifdef this, because of `FEATURE_EVENTSOURCE_XPLAT` is almost unused with this change. I do not have opinion what looks better. @manu-silicon What do you think?

**Comment by ramarag** [comment]
Ok, it makes sense from that point of view, although composition is lost. 
@manu-silicon : Can you remove all occurrences of EATURE_EVENTSOURCE_XPLAT, as it no longer has any meaning

**Comment by manu-silicon** [comment]
@jkotas @ramarag Looks like I'm receiving mix-messages here and I'm not entirely following the reasoning of keeping either or of FEATURE_PAL or FEATURE_EVENTSOURCE_XPLAT. The former is to distinguish between Unix and Windows and the later for Unix only to see if something is available or not.

**Comment by ramarag** [comment]
The reason behind is FEATURE_EVENTSOURCE_XPLAT is to compose functionality for logging EventSource Events distinctly from ETW period. This could potentially be reused on windows too [ not gonna happen- bu still saying]

FEATURE_PAL implies non windows to be exact

**Comment by jkotas** [comment]
Ok. @manu-silicon Could you please make these changes:

- Turn on FeatureXplatEventSource in clr.coreclr.props for all Unix (move it couple of lines down to the Unix block)
- Turn on DFEATURE_EVENTSOURCE_XPLAT=1 in CMakeLists.txt for all Unix
- Use FEATURE_EVENTSOURCE_XPLAT instead of FEATURE_PAL for ifdefing this

Sorry about the back and forth on this one...

**Comment by jkotas** [comment]
The ifdef for real implementation vs. the dummy implementation in `eventtracepriv.h` should be `#ifdef FEATURE_EVENT_TRACE`

**Comment by manu-silicon** [comment]
@jkotas Thanks for the clarification. I'll have a second pass at it tomorrow.

**Comment by ramarag** [comment]
by turning on FEATURE_EVENTSOURCE_XPLAT for all Linux, you do not need to have dummy implementation anywhere, this was the case when the feature got checked in, the current status is due to consolidation from this change https://github.com/dotnet/coreclr/commit/dd9c20ca70d650083e57802e5d1e9bfde799f9b9

**Comment by jkotas** [comment]
I think you still need to have the dummy implementation even when `FEATURE_EVENTSOURCE_XPLAT` is turned on for all Unix. Lttng implementation will be still only available only when `FEATURE_EVENT_TRACE` is defined.

**Comment by ramarag** [comment]
Almost all of the functionality defined by FEATURE_EVENT_TRACE   should have dummy default implementations already  if it is not defined.

**Comment by manu-silicon** [comment]
@jkotas @ramarag Here is my latest take on the matter. Compared to the original code, the only change is that now you can disable FEATURE_EVENTSOURCE_XPLAT when FEATURE_EVENT_TRACE is enabled, which was not possible before.

However, disabling FEATURE_EVENT_TRACE on Unix requires the set of changes I made originally if we do not want to change the way mscorlib is compiled. So I gave up doing that, which means that if someone disable it in CMakeLists.txt, it has to disable it in mscorlib too.

Let me know what you think.

File: src/vm/eventtracepriv.h [hunk 1]
```diff
@@ -408,13 +408,17 @@ class BulkStaticsLogger
 
 #endif // __EVENTTRACEPRIV_H__
 
-#if defined(FEATURE_EVENTSOURCE_XPLAT)
+#if defined(FEATURE_EVENT_TRACE)
```
Comment (jkotas) [inline src/vm/eventtracepriv.h]: I think this should stay `FEATURE_EVENTSOURCE_XPLAT`

File: src/vm/eventtracepriv.h [hunk 2]
```diff
@@ -408,13 +408,17 @@ class BulkStaticsLogger
...
 class XplatEventSourceLogger
 {
 public:
+#ifdef FEATURE_EVENTSOURCE_XPLAT
```
Comment (jkotas) [inline src/vm/eventtracepriv.h]: This should be `FEATURE_EVENT_TRACE`

File: src/vm/ecalllist.h [hunk 3]
```diff
@@ -2080,12 +2080,12 @@ FCFuncStart(gWindowsRuntimeBufferHelperFuncs)
 FCFuncEnd()
 #endif // ifdef FEATURE_COMINTEROP
 
-#if defined(FEATURE_EVENTSOURCE_XPLAT)
+#if defined(FEATURE_EVENT_TRACE)
```
Comment (jkotas) [inline src/vm/ecalllist.h]: Revert changes in this file

File: src/vm/mscorlib.cpp [hunk 4]
```diff
@@ -130,12 +130,12 @@
 #include "windowsruntimebufferhelper.h"
 #endif
 
-#if defined(FEATURE_EVENTSOURCE_XPLAT)
+#if defined(FEATURE_EVENT_TRACE)
```
Comment (jkotas) [inline src/vm/mscorlib.cpp]: Revert changes in this file

**Comment by jkotas** [comment]
`FEATURE_EVENTSOURCE_XPLAT` should control only whether the two internal calls from mscorlib to VM exists.

Please let me know if there is anything that I am missing...

**Comment by ramarag** [comment]
@jkotas : FEATURE_EVENTSOURCE_XPLAT controls the the all the infrstrcuture required to call  from mscorlib to vm and the Logging functionality in mscorlib.

Just a clarification,  the intent of this change is to
1. continue to NOT enable LTTNg for non ubuntu linux
2. Fix the qcall and fcall tables in vm due to mismatch from the entries in mscorlib and vm correct ?

If my second assumption is true ?, then revert all of the changes above add just add 
if(CLR_CMAKE_PLATFORM_UNIX)
add_definitions(-DFEATURE_EVENTSOURCE_XPLAT=1)
endif(CLR_CMAKE_PLATFORM_UNIX)

to  root/CMakeLists.txt

**Comment by jkotas** [comment]
If you make just the change you are proposing, you will get build errors (just tried it on my MacBook - see below). The changes that @manu-silicon made are required to fix these breaks.

```
Linking CXX shared library libcoreclr.dylib
Undefined symbols for architecture x86_64:
  "XplatEventSourceLogger::LogEventSource(int, char16_t const*, char16_t const*, char16_t const*)", referenced from:
      gEventLogger in libcee_wks.a(mscorlib.cpp.o)
  "XplatEventSourceLogger::IsEventSourceLoggingEnabled()", referenced from:
      gEventLogger in libcee_wks.a(mscorlib.cpp.o)
```

**Comment by ramarag** [comment]
That is because eventtrace.cpp is not included in the compile step:
https://github.com/dotnet/coreclr/blob/be3eab26a5017fde7f533cc22ae2c29ddfaa2aee/src/vm/CMakeLists.txt#L272

as CMake define "FEATURE_EVENT_TRACE" is not defined

**Comment by ramarag** [comment]
can you try if the changes from https://github.com/ramarag/coreclr/tree/xplateventsrc works ?

**Comment by jkotas** [comment]
@ramarag your delta changes would work with some more work, but I think @manu-silicon approach - that did not add vm\eventtrace.cpp to the compilation - is better. ifdefs that span almost the whole file are anti-pattern that is better to avoid.

**Comment by ramarag** [comment]
this can not be enabled without taking in vm\eventtrace.cpp, the only way to go around it to remove the code under FEATURE_EVENTSOURCE_XPLAT to another file

**Comment by leemgs** [comment]
@manu-silicon , It's strange. I can not build coreCLR with the below existing command for Linux/ARM after applying your PR https://github.com/dotnet/coreclr/pull/3879/commits/89d31c6c114fbd3414346be103db2eac9ccd9f65
Could you build the PR #3879 without any build breaks? Please, refer to the below information.

 * Here is a build command: 
`invain@u1404:/work/dotnet/arm-32/coreclr$ time ROOTFS_DIR=/work/dotnet/rootfs-coreclr/arm ./build.sh arm debug clean cross `

 * Below is the error messages in case of the build-break:
```
In file included from /work/dotnet/rootfs-coreclr/arm/usr/include/lttng/tracepoint-event.h:58:
In file included from /work/dotnet/rootfs-coreclr/arm/usr/include/lttng/ust-tracepoint-event.h:26:
In file included from /work/dotnet/rootfs-coreclr/arm/usr/include/urcu/rculist.h:30:
In file included from /work/dotnet/rootfs-coreclr/arm/usr/include/urcu-pointer.h:31:
In file included from /work/dotnet/rootfs-coreclr/arm/usr/include/urcu/uatomic.h:41:
/work/dotnet/rootfs-coreclr/arm/usr/include/urcu/uatomic/generic.h:75:10: error: address argument to atomic builtin must be a
      pointer to integer or pointer ('void *' invalid)
                return __sync_val_compare_and_swap_4(addr, old, _new);
                       ^                             ~~~~
/work/dotnet/rootfs-coreclr/arm/usr/include/urcu/uatomic/generic.h:112:3: error: address argument to atomic builtin must be a
      pointer to integer or pointer ('void *' invalid)
                __sync_and_and_fetch_4(addr, val);
                ^                      ~~~~
/work/dotnet/rootfs-coreclr/arm/usr/include/urcu/uatomic/generic.h:151:3: error: address argument to atomic builtin must be a
      pointer to integer or pointer ('void *' invalid)
                __sync_or_and_fetch_4(addr, val);
                ^                     ~~~~
/work/dotnet/rootfs-coreclr/arm/usr/include/urcu/uatomic/generic.h:190:10: error: address argument to atomic builtin must be
      a pointer to integer or pointer ('void *' invalid)
                return __sync_add_and_fetch_4(addr, val);
                       ^                      ~~~~
[  3%] Building C object src/pal/src/CMakeFiles/coreclrpal.dir/safecrt/safecrt_output_l.c.o
4 errors generated.
make[2]: *** [src/pal/src/eventprovider/tracepointprovider/CMakeFiles/coreclrtraceptprovider.dir/__/lttng/traceptprovdotnetruntime.cpp.o] \uc624\ub958 1
```

**Comment by manu-silicon** [comment]
This is strange, I thought this requirement was not needed but looks like we still need to patch urcu. Follow the instructions that were removed by PR #3658

**Comment by leemgs** [comment]
@manu-silicon , I have got the above error messages in case that I have only applied your PR #3879 .  Could you check a reason why your PR make urcu related error again like the previous build situation?

**Comment by manu-silicon** [comment]
Enabling FEATURE_EVENT_TRACE forces the inclusion of lttng which depends on urcu.

**Comment by jkotas** [comment]
@manu-silicon Could you please incorporate these comments and verify that it works on ARM?

https://github.com/dotnet/coreclr/pull/3879#discussion_r57578155
https://github.com/dotnet/coreclr/pull/3879#discussion_r57578193
https://github.com/dotnet/coreclr/pull/3879#discussion-r57578617
https://github.com/dotnet/coreclr/pull/3879#discussion-r57578664

Once these changes are made, I will merge this - assuming that everything works.

**Comment by manu-silicon** [comment]
@jkotas I'll see how it goes. I remember trying and that did not work as expected.

**Comment by manu-silicon** [comment]
@jkotas I've just done so and it does work. I think I was playing with enabling disabling flags. Note that as @leemgs discovered we need to add back the entry about urcu and cross platform builds. Locally I'm testing with FEATURE_EVENT_TRACE enabled and FEATURE_EVENTSOURCE_XPLAT disabled.

**Comment by jkotas** [comment]
I think we should keep `FEATURE_EVENT_TRACE` disabled for Linux ARM by default (to keep the overhead low by default, and to avoid the urcu issue - at least for now). Reverting the change in the root CMakeLists.txt should do the trick.

**Comment by manu-silicon** [comment]
FEATURE_EVENT_TRACE enabled and FEATURE_EVENTSOURCE_XPLAT disabled works.

Disabling FEATURE_EVENT_TRACE now for ARM to see if this will work.

**Comment by manu-silicon** [comment]
Seems to work for me. @myungjoo @leemgs Could you try this to see if this addresses all issues?

**Comment by myungjoo** [comment]
@prajwal-aithal , Could you please test this one? (Me and @leemgs are away for a week and cannot test ARM devices)

**Comment by jkotas** [comment]
The OSX build is failing. The problem is that FEATURE_EVENTSOURCE_XPLAT has to be defined for Unix unconditionally in the root CMakefile.txt.

**Comment by manu-silicon** [comment]
You mean regardless of the value of FEATURE_EVENT_TRACE?

**Comment by manu-silicon** [comment]
Do you mean having something like that in the CMakeLists.txt
```
if(CLR_CMAKE_PLATFORM_LINUX AND CLR_CMAKE_PLATFORM_ARCH_AMD64)
  set(FEATURE_EVENT_TRACE 1)
endif()
if(CLR_CMAKE_PLATFORM_UNIX)
  add_definitions(-DFEATURE_EVENTSOURCE_XPLAT=1)
endif()
```
?

I just tried on my Mac and it works. I did not test if that would work for arm but will try tomorrow.

**Comment by jkotas** [comment]
Yes, exactly that.

**Comment by manu-silicon** [comment]
I pushed the changes. It works for me on Mac and Linux ARM. Did not test anything else.

**Comment by ramarag** [comment]
If it is not too much too ask can you build on top of https://github.com/ramarag/coreclr/commit/84a8d4d375dea0c59eabbae11e3e68d9532e23f1 ?

**Comment by jkotas** [comment]
@ramarag I would like to get this done in steps. Make one small step to get it working - because of there are people waiting for this, and then work on better factoring. The implementation of the event tracing is very entangled and the history of this PR shows that it is hard to do bigger steps here (e.g. there are still build breaks).

File: src/vm/eventtracepriv.h [hunk 5]
```diff
@@ -412,9 +412,13 @@ class BulkStaticsLogger
 class XplatEventSourceLogger
 {
 public:
+#ifdef FEATURE_EVENT_TRACE
```
Comment (ramarag) [inline src/vm/eventtracepriv.h]: This conditional compilation is done for FireEtw* and XplatEventLogger::IsEventLoggingEnabled under the hoods. If it is not pressing to check in the change as it is please do look to build on top of https://github.com/ramarag/coreclr/commit/84a8d4d375dea0c59eabbae11e3e68d9532e23f1

Comment (manu-silicon) [inline src/vm/eventtracepriv.h]: Do you mean compiling just your commit? or  Mine and yours combined?

Comment (manu-silicon) [inline src/vm/eventtracepriv.h]: With your commit, it seems to work.

**Comment by jkotas** [comment]
@manu-silicon The global CMakefile is sensitive to order in which things are defined. You have moved them earlier and it broke the build. Could you please keep it where it was, and just tweak the ifdef?

**Comment by manu-silicon** [comment]
@jkotas I wanted to put them together. I've pushed the change

File: src/vm/eventtracepriv.h (hunk 5 above)
Comment (ramarag) [inline src/vm/eventtracepriv.h]: yes i meant add your changes on top of ramarag@84a8d4d

**Comment by jkotas** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test  please

**Comment by jkotas** [comment]
@manu-silicon I am sorry - I have just merged a number of other PRs while the CI was running on this one, and it has conflict now. Could you please resolve it?

**Comment by jkotas** [comment]
@ramarag I have created issue #4403 about the cleanup that you are suggesting.

**Comment by manu-silicon** [comment]
@jkotas Just pushed the changed.

**Comment by jkotas** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test please

**Comment by jkotas** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test please

**Comment by jkotas** [comment]
@manu-silicon Thank you for all work you have put into this one!

**Comment by manu-silicon** [comment]
@jkotas Not much work, but it was definitely time consuming. Glad we can now on.
@ramarag I can try to test your code, but could you update your branch to the latest master to make the process easier on me. Thanks!

**Comment by ramarag** [comment]
@manu-silicon : i have updated the changes to https://github.com/ramarag/coreclr/commit/0ec237e25da01a855c6ed2ad77d2500f7d3aa34b

**Pull request merged**